add Haptic Sense Glove Wearable implementation

kouroshD opened the pull request.

diegoferigo commented on cmake/FindSenseGlove.cmake:
```diff
@@ -0,0 +1,34 @@
+# Copyright 2020 Istituto Italiano di Tecnologia (IIT)
```
2021?

diegoferigo commented on cmake/FindSenseGlove.cmake:
```diff
@@ -0,0 +1,34 @@
+target_include_directories(SenseGlove INTERFACE $ENV{SenseGlove_INCLUDE_DIRS})
+
+set(SenseGlove_FOUND TRUE)
+
+
+
```
cc @traversaro, I'm not expert in developing `Find*.cmake` files.

traversaro replied:
It seems ok, the only question is if there is some docs somewhere on how the SenseGlove SDK needs to be installed and used? Because the only thing different from the usual `Find**.cmake`  workflow is that the `CMAKE_PREFIX_PATH`  is not searched for libraries and headers files, but rather `SenseGlove_DIR` env variable **must** be defined.

lrapetti replied:
>  if there is some docs somewhere on how the SenseGlove SDK needs to be installed and used?

Regarding this point, a minimum information should be added in the `README.md` at https://github.com/robotology/wearables#wearable-device-sources

kouroshD replied:
> It seems ok, the only question is if there is some docs somewhere on how the SenseGlove SDK needs to be installed and used? Because the only thing different from the usual Find**.cmake workflow is that the CMAKE_PREFIX_PATH is not searched for libraries and headers files, but rather SenseGlove_DIR env variable must be defined.

What I was thinking to do, is later fork this repo https://github.com/Adjuvo/SenseGlove-API to DIC organization. Later, there are two environment variables to set.

kouroshD replied:
@traversaro  Note that, I am using currently two environment variables for the senseGlove, i.e., `SenseGlove_DIR` and `SenseGlove_INCLUDE_DIRS`. Moreover, `SenseGlove_DIR` is passed to `PATH` as well. If this is OK, I resolve this conversation.

traversaro replied:
Can't we put this info in the README as suggested by @lrapetti ? If this takes time no problem, but if it is fast is probably worth to do it!

kouroshD replied:
I am adding a README to the HapticGlove Floder folder

traversaro replied:
Thanks! Probably it could make sense to link it from the main README, otherwise it may be difficult to locate.

kouroshD replied:
I have added the README to the device and also to the https://github.com/robotology/wearables#wearable-device-sources

Let me know if it is OK.

lrapetti replied:
Thanks, @kouroshD. Just a small comment, in order to be coherent, maybe you could move it in [`doc`](https://github.com/robotology/wearables/tree/master/doc) together with the documentation of the other devices. Eventually you can call it `how-to-compile-and-run...`

kouroshD replied:
Can I resolve this conversation, I have updated with the requested changes?

traversaro replied:
Ok for me.

diegoferigo commented on devices/HapticGlove/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+#/*
+# Copyright 2021 Istituto Italiano di Tecnologia (IIT)
+# @author Kourosh Darvish <kourosh.darvish@iit.it>
+#* Authors: Yeshasvi Tirupachuri
+#* CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+#*/
```
Do we have a standardized header?

kouroshD replied:
I am not sure, updating the file.

diegoferigo commented on devices/HapticGlove/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+#* CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+#*/
+
+find_package(Eigen3 REQUIRED)
+include_directories(SYSTEM ${EIGEN3_INCLUDE_DIR})
```
I think that the `NO_MODULE` version of Eigen3 does not require using `include_directories`, just link against `Eigen3::Eigen`.

```suggestion
find_package(Eigen3 REQUIRED NO_MODULE)
```

diegoferigo commented on devices/HapticGlove/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+    YARP::YARP_dev
+    YARP::YARP_init
+    Wearable::IWear
+    SenseGlove
```
```suggestion
    PRIVATE
    SenseGlove
    Eigen3::Eigen
```

diegoferigo commented on devices/HapticGlove/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+set (WEARABLES_PROJECT_NAME Wearables)
+set (WEARABLES_XML_FILES conf/HapticGlove.xml)
+install(FILES ${WEARABLES_XML_FILES}
+            DESTINATION ${CMAKE_INSTALL_DATADIR}/${WEARABLES_PROJECT_NAME})
```
```suggestion
        DESTINATION ${CMAKE_INSTALL_DATADIR}/${WEARABLES_PROJECT_NAME})
```

diegoferigo commented on devices/HapticGlove/include/HapticGlove.h:
```diff
@@ -0,0 +1,252 @@
+ * @date 2021
+ */
+
+#ifndef HAPTIC_GLOVE_H
+#define HAPTIC_GLOVE_H
```
Can you please format the entire PR with the shipped clang-format file?

Yeshasvitvs replied:
@kouroshD References for using [clang-format file](https://github.com/robotology/wearables/blob/master/.clang-format) : https://www.vikingsoftware.com/using-clang-format-with-qtcreator/

kouroshD replied:
Done, added.

diegoferigo commented on devices/HapticGlove/include/HapticGlove.h:
```diff
@@ -0,0 +1,252 @@
+    inline ElementPtr<const actuator::IMotor>
+    getMotorActuator(const actuator::ActuatorName) const override;
+
+// TODEL
+//    inline SensorPtr<const sensor::IOrientationSensor>
+//    getOrientationSensor(const sensor::SensorName /*name*/) const override;
+
+//    bool configure (const yarp::os::Searchable& config, const std::string& name);
+
+//    bool update(const std::vector<double>& imuData);
```
leftover?

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+    // Numbe of actuators
+    const int nActuators = 10; // humanFingerNames.size()*2
+
+    std::unique_ptr<senseGlove::SenseGloveHelper> m_glove; /**< Pointer to the glove object. */
```
Can you please uniform the names of the private attributes following the style:

1. If an attribute is declared in the public header of the device, use the `m_` prefix.
2. If an attribute is declared in the implementation class, don't use `m_` since it's redundant. The main class will use `pImpl->privAttr` to access it. Use `this->privAttr` in the Impl methods definition to make it clear that is a class attribute.

kouroshD replied:
`this->privAttr` but this is not obligatory, right? I can use the attribute without the `this` pointer , e.g., `privAttr`?

diegoferigo replied:
Nope, it's just for readability. Using `pImpl->` or `this->` makes clear that it's a class attribute. If I don't see any, I always look for a local variable without finding it. It's not a must, it just improves readability.

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+{
+    std::lock_guard<std::mutex> lock(mutex);
+
+    wearableName="SenseGlove";
+    portsPrefix ="/wearable/SenseGlove";
+    m_glove= std::make_unique<senseGlove::SenseGloveHelper>();
```
Move these in the members initializers list?

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+
+HapticGlove::HapticGlove()
+    : PeriodicThread(period)
+    , pImpl{new SenseGloveImpl()}
```
```suggestion
    , pImpl{std::make_unique<SenseGloveImpl>()}
```

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+    // ===================
+    // Ports configuration ???
+    // ===================
+//    if(!pImpl->dataPort.open(pImpl->portsPrefix + ":o")) {
+//        yError() << LogPrefix << "Failed to open data port " << pImpl->portsPrefix + ":o";
+//        return false;
+//    }
```
leftover?

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+//    }
+
+    pImpl->linkSensorPrefix =getWearableName() +sensor::IVirtualLinkKinSensor::getPrefix();
+    pImpl->sensegloveLinkSensor= SensorPtr<SenseGloveImpl::SenseGloveVirtualLinkKinSensor>{std::make_shared<SenseGloveImpl::SenseGloveVirtualLinkKinSensor>(pImpl.get(),
+                                                                                                                                                           pImpl->linkSensorPrefix + pImpl->gloveData.humanHandLinkName)};
```
Is it necessary to use also `SensorPtr` here beyond the `make_shared` call?

kouroshD replied:
About this point I am not sure, I tried to follow other implementations.

diegoferigo replied:
Blindly, this could maybe work?

```cpp
pImpl->sensegloveLinkSensor = std::make_shared<SenseGloveImpl::SenseGloveVirtualLinkKinSensor>(
    pImpl.get(), pImpl->linkSensorPrefix + pImpl->gloveData.humanHandLinkName);
```

kouroshD replied:
As you suggested, and I agree with you, they are redundant initialization. I removed them from the related parts of the code.

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+
+    bool getLinkAcceleration(Vector3& linear, Vector3& angular) const override
+    {
+        // we do not handle linear and angular accelerations in the current implementation
```
maybe printing a warning in case someone will call this method by mistake?

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,621 @@
+
+        assert(m_gloveImpl != nullptr);
+
+        std::lock_guard<std::mutex> lock(m_gloveImpl->mutex);
```
Should this mutex also protect the position? Even if it is not currently handled. Also, I'd use the mutex is all `get*` methods of all sensor implementations.

kouroshD replied:
done.

diegoferigo reviewed (changes requested):
First quick pass, I will do another one more detailed.

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,637 @@
+
+    bool getLinkAcceleration(Vector3& linear, Vector3& angular) const override
+    {
+        assert(m_gloveImpl != nullptr);
```
Maybe this check can be done just once in the constructor?

diegoferigo commented on devices/HapticGlove/test/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+
+install(TARGETS ${PROJECT_NAME} DESTINATION bin)
+
+install(FILES config.ini DESTINATION bin )
```
Are we sure we want to have a `bin/config.ini`? I suspect this might create a lot of problems because it would appear in the command autocompletion of any terminal.

kouroshD replied:
I added here, since I had in mind this code as a test code and not installing it inside superbuild. But since, now, it can be used to visualize the frames of any wearable data/device, I was thinking to move this snippet as a module to `wearable/modules/IWearFrameVisualizer`. What do you think?
cc @Yeshasvitvs @lrapetti

diegoferigo replied:
Yep! I'm not sure where these configs are typically installed by wearables, I expect somewhere in `CMAKE_INSTALL_DATADIR` (share/).

kouroshD replied:
I have updated the it as we decided. I added the `modules` folder, then I installed it on the directory you suggested correctly. Let me know if this is OK.

diegoferigo replied:
Looks good!

diegoferigo commented on devices/HapticGlove/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+target_link_libraries(HapticGlove
+    PUBLIC
+    YARP::YARP_dev
+    YARP::YARP_init
```
I think that `YARP::YARP_init` could be private

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,711 @@
+
+const std::string DeviceName = "HapticGlove";
+const std::string LogPrefix = DeviceName + wearable::Separator;
+double period = 0.01;
```
Is this the default period or it can also be configured?

kouroshD replied:
This is the default value, later it is updated witht the value read from the xml file.

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,711 @@
+  ~SenseGloveVirtualLinkKinSensor() override = default;
+
+  bool getLinkAcceleration(Vector3 &linear, Vector3 &angular) const override {
+    assert(m_gloveImpl != nullptr);
```
Maybe this check can only be done once in the constructor

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,711 @@
+  ~SenseGloveVirtualJointKinSensor() override = default;
+
+  bool getJointPosition(double &position) const override {
+    assert(m_gloveImpl != nullptr);
```
Same for this class

diegoferigo commented on devices/HapticGlove/src/HapticGlove.cpp:
```diff
@@ -0,0 +1,711 @@
+    return nullptr;
+  }
+
+  return dynamic_cast<
```
Just checking, is the `dynamic_cast` necessary? Downcast should be done automatically. If it is necessary, and your sure that the original data can be successfully cast, I'd recommend using `static_cast` that has less runtime cost.

kouroshD replied:
changed them to static cast.

diegoferigo reviewed (commented):
other quick pass

Yeshasvitvs commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,281 @@
+namespace senseGlove {
+class SenseGloveHelper;
+const std::string LogPrefix = "senseGlove::SenseGloveHelper::";
+const size_t PoseSize = 7;
```
@kouroshD Please add any reference that is related to the value `7`

kouroshD replied:
I added a comment for this. `// [ position <x, y, z>, quaternion <w, x, y, z> ]`

kouroshD replied:
let me know if this is OK

Yeshasvitvs commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,281 @@
+
+class senseGlove::SenseGloveHelper {
+
+  int m_forceFbDof; /**< Number of the actuated motors Dofs to produce force
```
May be using [variable comment](https://google.github.io/styleguide/cppguide.html#Variable_Comments) style will be cleaner

kouroshD replied:
I added this style for the comments, if we decide to make [doxygen](https://www.doxygen.nl/manual/docblocks.html) we have the comments for the members and methods.

diegoferigo replied:
If lines get too long after you apply clang format, you can move the comment before the definition (and adapt properly the special comment syntax)

kouroshD replied:
As you suggested, I added the comments before the member variables.

kouroshD replied:
I updated this, let me know if this is OK.

Yeshasvitvs commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,281 @@
+
+  /**
+   * Configure the class
+   * @param config confifuration options
```
Typo `confifuration` -------> `configuration`

Yeshasvitvs commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,281 @@
+   * get the number of buzz motors/vibro-tactile feedback motors
+   * @return number of buzz motors
+   */
+  int getNoOfBuzzMotors() const;
```
using `Num` instead of `No` for the functions associated with getting number of elements ?

Yeshasvitvs commented on devices/HapticGlove/test/VisualizeFrames.cpp:
```diff
@@ -0,0 +1,355 @@
+#endif
+}
+
+struct FrameViewier {
```
Typo `FrameViewier` ---------> `FrameViewer`

Yeshasvitvs commented on devices/HapticGlove/test/VisualizeFrames.cpp:
```diff
@@ -0,0 +1,355 @@
+
+  // Iwear Remapper
+  yarp::os::Value *wearableDataPort;
+  if (!config.check("wearable_data_ports", wearableDataPort)) {
```
If the config parameter is a single port, better to have the associate variable also singular `wearable_data_port`

Yeshasvitvs replied:
NVM, I see the config file has wearable data ports from left and right. So, may be `wearableDataPorts`

kouroshD replied:
Consider that, here https://github.com/robotology/wearables/blob/c47205dae6c6c508f2f39545e2a7e835d3bf3610/devices/IWearRemapper/src/IWearRemapper.cpp#L129-L140 it accepts a list of strings not a string!

kouroshD replied:
@Yeshasvitvs  Can I resolve this conversation?

diegoferigo commented on cmake/FindSenseGlove.cmake:
```diff
@@ -0,0 +1,34 @@
+
+# Check Directory of the SenseGlove_DIR
+if(NOT DEFINED ENV{SenseGlove_DIR})
+  message( FATAL_ERROR "Environment variable {SenseGlove_DIR} is not defined: $ENV{SenseGlove_DIR}" )
```
what would be printed in the end of the error if `$ENV{SenseGlove_DIR}` is not defined?

kouroshD replied:
probably an empty string! I do not remember! I can remove it!

diegoferigo replied:
Ok, also below

kouroshD replied:
I removed it.

diegoferigo commented on cmake/FindSenseGlove.cmake:
```diff
@@ -0,0 +1,34 @@
+if(NOT DEFINED ENV{SenseGlove_DIR})
+  message( FATAL_ERROR "Environment variable {SenseGlove_DIR} is not defined: $ENV{SenseGlove_DIR}" )
+else()
+  message("Environment variable {SenseGlove_DIR}: $ENV{SenseGlove_DIR}" )
```
message(STATUS ...)?

Also below

diegoferigo commented on cmake/FindSenseGlove.cmake:
```diff
@@ -0,0 +1,34 @@
+  message("Environment variable {SenseGlove_DIR}: $ENV{SenseGlove_DIR}" )
+endif()
+
+if(NOT DEFINED ENV{SenseGlove_INCLUDE_DIRS})
```
Could this be inferred from `$ENV{SenseGlove_DIR}`?

kouroshD replied:
Here, `SenseGlove_DIR` is the directory that we have the libraries; this is passed to `PATH` as well.
So it is not the root of the git repo related to senseGlove. If I want to infer it, I sould add sth like `../../<path to header files>`.
Then if I want to add `SenseGlove_DIR` as the root of the repo, both for the `PATH` environment variable and inside the cmake file, I should add the folder. 
What is your suggestion?

cc @traversaro  @S-Dafarra

kouroshD replied:
An option might be also, to clone the senseGlove repo with superbuild, if the option is enabled.

S-Dafarra replied:
I think I am missing something. Is the relative path between ``SenseGlove_DIR`` and ``SenseGlove_INCLUDE_DIRS`` always fixed?

Also, if they have a common parent, it may also make sense to define that parent as ``SenseGlove_DIR``. For example, assume that the structure is as follows
```
- Senseglove_SDK_DIR
              |
              |-- bin
                    |
                    |-- someLib.dll
              |
              |-- include
                      |
                      |-- Senseglove
                                    |
                                    |-- SenseGlove.h
```
you could ask the user to set ``SenseGlove_DIR`` equal to the path of the Senseglove_SDK_DIR, then you hardcode the path to ``someLib.dll`` and ``SenseGlove.h`` starting from the ``SenseGlove_DIR``.

kouroshD replied:
I have updated the cmake file. Let me know if it looks fine.

diegoferigo replied:
@S-Dafarra that is exactly what I meant, looks much better now!

kouroshD replied:
Can I resolve this conversation, I have updated the cmake files.

diegoferigo commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,250 @@
+
+class senseGlove::SenseGloveHelper
+{
+    int m_forceFbDof; /**< Number of the actuated motors Dofs to produce force feedback to the human*/
```
What I mean is the following:

```suggestion
    /// Number of the actuated motors Dofs to produce force feedback to the human
    int m_forceFbDof;
```

Otherwise, clang-format will mess everything up as soon as you will process all PR sources.

kouroshD commented:
@DanielePucci  Can I fork this repo https://github.com/Adjuvo/SenseGlove-API to DIC organization? Since, they may change the code time to time, and we may break our code because of dependencies.

traversaro commented:
> 
> 
> @DanielePucci Can I fork this repo https://github.com/Adjuvo/SenseGlove-API to DIC organization? Since, they may change the code time to time, and we may break our code because of dependencies.

If we don't need any modification, we can just document the specific commit of that repo that we use, i.e. in the docs have:
> Clone the repo at the specified commit:
~~~
git clone https://github.com/Adjuvo/SenseGlove-API
cd SenseGlove-API
git checkout baad587d4e165d7bfafd7f0c8ee6a1ce8ad651d6
~~~

However, I just forked the repo in the case you ever need it.

traversaro commented:
> 
> 
> > @DanielePucci Can I fork this repo https://github.com/Adjuvo/SenseGlove-API to DIC organization? Since, they may change the code time to time, and we may break our code because of dependencies.
> 
> If we don't need any modification, we can just document the specific commit of that repo that we use, i.e. in the docs have:
> 
> > Clone the repo at the specified commit:
> 
> ```
> git clone https://github.com/Adjuvo/SenseGlove-API
> cd SenseGlove-API
> git checkout baad587d4e165d7bfafd7f0c8ee6a1ce8ad651d6
> ```
> 
> However, I just forked the repo and gave you rights on in the case you ever need it: https://github.com/dic-iit/SenseGlove-API .

I opened https://github.com/Adjuvo/SenseGlove-API/issues/9 to ask if they could tag their release, so that the code could be simplified to:
~~~
git clone https://github.com/Adjuvo/SenseGlove-API
cd SenseGlove-API
git checkout v1.3
~~~

kouroshD commented:
I have tested both Haptic Glove device and Frame Visualizer Application on Windows. They are working correctly.

lrapetti commented on README.md:
```diff
@@ -41,3 +41,4 @@ Wearable devices contained in [`/devices/`](/devices) are [YARP devices](http://
 | XsensSuit | Exposes [XsensSuit](https://www.xsens.com/motion-capture) data. |  Windows   | xsens MVN SDK 2018.0.3 | [:books:](/doc/How-to-run-XsensSuit.md) |
 | iCub | Exposes [iCub](https://icub.iit.it/) robot data. |  Linux/MacOS/Windows   | [idyntree](https://github.com/robotology/idyntree) | [:books:](/doc/How-to-run-iCub-as-wearable-source.md) |
 | Paexo | Exposes [Paexo](https://paexo.com/?lang=en) data. |  Linux  | Optional flag `ENABLE_PAEXO_USE_iFEELDriver` to use `iFeelDriver` (Contact the maintainer for more details) | - |
+| HapticGlove | Exposes [SenseGlove](https://www.senseglove.com/product/developers-kit/) data. |  Linux/Windows  | [SenseGloveSDK](https://github.com/Adjuvo/SenseGlove-API) | [HapticGloveDocumentation](./devices/HapticGlove/README.md) |
```
For the other devices, we have used the emoji 📚  in place of the name of the device for the documentation link. You can use the same for the gloves (`[:books:](/path/to/the/documentation.md)`).

lrapetti reviewed (approved):
I added just some minor comments on the documentation, and the `CHANGELOG` still needs to be updated.
Rather than that, if all the other comments have been addressed, for me we can proceed.

diegoferigo commented:
LGTM! Just remember to format all the files with the right clang-format configuration (we should add a github action to enfore it in the CI level).

kouroshD commented:
> LGTM! Just remember to format all the files with the right clang-format configuration (we should add a github action to enfore it in the CI level).

@diegoferigo  I have installed the clang-format, and all the files for this PR should be with the clang format. Let me know if it sounds OK, or needs some adjustments!

kouroshD commented:
@Yeshasvitvs  Do you have other comments on this PR, or we can proceed with the merging?

kouroshD commented:
> I added just some minor comments on the documentation, and the CHANGELOG still needs to be updated.
Rather than that, if all the other comments have been addressed, for me we can proceed.

Updated with the requested changes.

diegoferigo commented on devices/HapticGlove/include/HapticGlove.h:
```diff
@@ -0,0 +1,244 @@
+
+    ElementPtr<const actuator::IActuator> getActuator(const actuator::ActuatorName name) const override;
+
+    VectorOfElementPtr<const actuator::IActuator> getActuators(const actuator::ActuatorType type) const override;
```
@kouroshD this line stills seems too long to me, did you process this file with clang format?

kouroshD replied:
I have made a dummy change in the file, and saved it. So, now it should be consistent with the clang format.

diegoferigo commented on devices/HapticGlove/include/SenseGloveHelper.hpp:
```diff
@@ -0,0 +1,326 @@
+  Object_Grasp_30 = 9
+};
+
+class senseGlove::SenseGloveHelper {
```
@kouroshD also, if I recall this bracket should go in a new line with the current configuration (top-level folder of the repo), make sure your clang-format picks it up

kouroshD replied:
I have tries to move the bracket in a new line, and it changed to the same line as you can see here. So, I would say that is the clang-format. Let me know if I should do, sth else?

Yeshasvitvs commented:
I agree with @diegoferigo the code is not updated with clang-format of wearables, for reference of how clang-formatting looks please check https://github.com/robotology/wearables/pull/123/commits/535075cff3c6cc39338c78649ebc225eed52689d

kouroshD commented:
>I agree with @diegoferigo the code is not updated with clang-format of wearables, for reference of how clang-formatting looks please check 535075c

I checked all the files, made some dummy changes to be with the clang-format. If some files are not still with the clang-format, please let me know.

diegoferigo commented:
> > I agree with @diegoferigo the code is not updated with clang-format of wearables, for reference of how clang-formatting looks please check [535075c](https://github.com/robotology/wearables/commit/535075cff3c6cc39338c78649ebc225eed52689d)
> 
> I checked all the files, made some dummy changes to be with the clang-format. If some files are not still with the clang-format, please let me know.

The problem seems that is it not picking up the configuration file of this project, falling back to a default style that is different to what we use.

On the fly I don't remember how to use the command line, my QtCreator automatically detects the `.clang-format` file in the top-level and formats all the open files when I (compulsory) press save while coding. Which IDE are you using?

kouroshD commented:
> On the fly I don't remember how to use the command line, my QtCreator automatically detects the .clang-format file in the top-level and formats all the open files when I (compulsory) press save while coding. Which IDE are you using?

![Screenshot from 2021-07-12 16-20-13](https://user-images.githubusercontent.com/17707730/125303514-2775d800-e32d-11eb-9919-3171bb735875.png)

I am using qtcreator, and it is recognizing the clang-format. When I do `save` a file, it will do compulsory formatting according to the clang-format!

diegoferigo commented:
You need to set it to `file`

![Screenshot_20210712_162620](https://user-images.githubusercontent.com/469199/125304560-fea21280-e32d-11eb-8314-18e20c10f687.png)

traversaro commented:
> I opened [Adjuvo/SenseGlove-API#9](https://github.com/Adjuvo/SenseGlove-API/issues/9) to ask if they could tag their release, so that the code could be simplified to:

Thanks to the awesome reply on SenseGlove side, we now have a v1.3.1 release tag: https://github.com/Adjuvo/SenseGlove-API/issues/9#issuecomment-878323477 .

kouroshD commented:
> You need to set it to `file`

Thank you, I have updated them as you mentioned. Let me know, if now seems to be OK on your eyes.

kouroshD commented:
Let's wait for the CI, and I will merge it.

Thank you all for your review.

diegoferigo reviewed (approved):
LGTM! Thanks @kouroshD

lrapetti commented:
@kouroshD check if you want to clean the commit history, otherwise we can proceed squashing and merging

kouroshD commented:
> @kouroshD check if you want to clean the commit history, otherwise we can proceed squashing and merging

I have updated the commit history, so better to proceed with normal merging.

kouroshD commented:
@Yeshasvitvs  @lrapetti  Can I proceed with the merge, or you prefer to do or a final check?
 We should do a normal merge, without squashing, since the history is clean now.

Yeshasvitvs commented:
> @Yeshasvitvs @lrapetti Can I proceed with the merge, or you prefer to do or a final check?

Fine by me 👍🏾
